### PR TITLE
Change PLC polling algorithm

### PR DESCRIPTION
**Pull request by ggevorg**

### CODE REVIEW

File: handler/plchandler/plchandler.go [hunk 1]
```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
...
-						fmt.Printf("error: %v\n", err)
-					}
+			//fmt.Printf("have command: %v\n", command)
+			tag, ok := ph.tags[command.Name]
+			if ok {
+				var val uint16 = on
+				if command.Value == "0" {
+					val = off
+				}
+				_, err := (*client).WriteSingleCoil(uint16(tag.Address), val)
```
Comment (aborilov) [inline handler/plchandler/plchandler.go]:
```
				if _, err := (*client).WriteSingleCoil(uint16(tag.Address), val); err != nil {
					fmt.Printf("error: %v\n", err)
				}

```

File: handler/plchandler/plchandler.go [hunk 2]
```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
...
+	sort.Slice(tagsMemBlocks, func(i, j int) bool {
+		return tagsMemBlocks[i].address < tagsMemBlocks[j].address
+	})
+
+	block := MemBlock{tags: []*Tag{}}
+	for _, mb := range tagsMemBlocks {
+		if block.size == 0 {
+			block.address = mb.address
+			block.size = mb.size
+			block.tags = append(block.tags, mb.tags[0])
```
Comment (aborilov) [inline handler/plchandler/plchandler.go:97]: is mb.tags can be empty?

File: handler/plchandler/plchandler.go (hunk 1 above)
Comment (aborilov) [reply handler/plchandler/plchandler.go]: (*client). why?

File: handler/plchandler/plchandler.go [hunk 3]
```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
...
-	defer ph.ClientHandler.Close()
-	//
-	client := modbus.NewClient(ph.ClientHandler)
+	defer ph.clientHandler.Close()
+
+	client := modbus.NewClient(ph.clientHandler)
+	return ph.loop(&client)
+}
+
+func (ph *PLCHandler) loop(client *modbus.Client) error {
```
Comment (aborilov) [inline handler/plchandler/plchandler.go]: modbus.Client is interface, you should not use pointer to interface

File: handler/plchandler/plchandler.go [hunk 4]
```diff
@@ -54,84 +71,125 @@ func NewPLCHandler(core handler.CoreHandler) *PLCHandler {
...
+				_, err := (*client).WriteSingleCoil(uint16(tag.Address), val)
+				if err != nil {
+					fmt.Printf("error: %v\n", err)
 				}
 			}
 		default:
-			for _, tag := range ph.Tags {
-				results, err := client.ReadCoils(uint16(tag.Address), 1)
+			for _, mb := range ph.pollingMemBlocks {
+				results, err := (*client).ReadCoils(uint16(mb.address), uint16(mb.size))
```
Comment (aborilov) [inline handler/plchandler/plchandler.go]: not need for pointer to client

File: handler/plchandler/plchandler.go [hunk 5]
```diff
@@ -13,19 +14,34 @@ import (
...
+
+	typeInput  = "input"
+	typeOutput = "output"
+
+	maxBlockSize = 64
+	maxBreakSize = 8
+)
+
+type MemBlock struct {
+	address int
```
Comment (aborilov) [inline handler/plchandler/plchandler.go]: let's make it uint16

File: handler/plchandler/plchandler.go (hunk 4 above)
Comment (aborilov) [inline handler/plchandler/plchandler.go]: no need for all uint16 conversation if we change address and size to uint16 in a first place

File: handler/plchandler/plchandler.go (hunk 1 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go]: fixed

File: handler/plchandler/plchandler.go (hunk 2 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go:97]: no

File: handler/plchandler/plchandler.go (hunk 3 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go]: fixed

File: handler/plchandler/plchandler.go (hunk 4 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go]: fixed

File: handler/plchandler/plchandler.go (hunk 5 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go]: fixed

File: handler/plchandler/plchandler.go (hunk 4 above)
Comment (ggevorg) [reply handler/plchandler/plchandler.go]: fixed

**Comment by aborilov** [comment]
LGMT

**Pull request merged**